### DISCUSSION TITLE
BRouter no-go areas in Cruiser

### DISCUSSION THREAD

**Discussion by WolfgangZ42** [devemux86/cruiser#85] [Q&A] [answered]
Trying to handle the traffic ban for "old" Diesel cars in Munich I'm thinking about a nogo area in BRouter. 
Reading and playing around ... but no success 😒 
I put a question in a forum:  https://groups.google.com/g/osm-android-bikerouting/c/5a7VldSPlAo/m/2lIEQPCNCQAJ  

But perhaps some of the Cruiser fans has also an idea ... 

To illustrate my plan:

<img width="400" src="https://user-images.githubusercontent.com/121806238/225591996-89dc5d1b-e174-4ea4-b043-519588a1e840.jpg"/>

**Comment by devemux86** [comment]
@WolfgangZ42 thanks for the question.

Cruiser generates and uses nogo areas in navigation with all supported routing services.
e.g. avoid roadblock

I plan to add a user interface to set nogo areas in the future.

---

About BRouter app, indeed their suggestion now is:
- https://github.com/abrensch/brouter/issues/312#issuecomment-905270043

Have you asked in [BRouter repository](https://github.com/abrensch/brouter/)?

For reference (are their docs up to date?):
- [BRouter documentation](https://brouter.de/brouter/vianogo.html)
- [BRouter readme](https://brouter.de/brouter/readme.txt)
- https://github.com/abrensch/brouter/issues/444
- https://github.com/abrensch/brouter/issues/429

**Comment by devemux86** [comment]
There is confirmation of the report in [BRouter forum](https://groups.google.com/g/osm-android-bikerouting/c/5a7VldSPlAo/m/2lIEQPCNCQAJ) and an issue was created by @afischerdev in BRouter repository:
- https://github.com/abrensch/brouter/issues/512

**Reply by WolfgangZ42** [reply]
Yes - I just saw it! Thanks afischerdev ... 
And believe me, I read the items you mentioned above. 
Thanks for your help!

**Comment by WolfgangZ42** [comment]
The nogo waypoint still doesn't work ... although there is a fix (https://github.com/abrensch/brouter/pull/513) ...
But: I got an interims solution - see https://groups.google.com/g/osm-android-bikerouting/c/5a7VldSPlAo/m/n3YwQTmbCQAJ ... and this works in Cruiser!
I'll keep you informed!

**Reply by devemux86** [reply]
https://github.com/abrensch/brouter/pull/513 contains a code change, so we need a new BRouter (Beta) apk to work.

I will inform when @afischerdev creates a new BRouter (Beta) apk:
- https://github.com/devemux86/cruiser/discussions/15

**Comment by devemux86** [comment]
@WolfgangZ42 and everyone:

@afischerdev has posted some questions about the relation of beelines and nogo areas:
- https://github.com/abrensch/brouter/issues/512#issuecomment-1474162809
- [BRouter forum](https://groups.google.com/g/osm-android-bikerouting/c/5a7VldSPlAo/m/rhm0ThfgAAAJ)

Can you check them and write your opinions there?

**Reply by WolfgangZ42** [reply]
Sure - but not until tomorrow, today I'm planned.

**Reply by WolfgangZ42** [reply]
I've answered one of the open items: https://groups.google.com/g/osm-android-bikerouting/c/5a7VldSPlAo/m/QfQu7mn8AAAJ 
I'll think about the other questions and discuss them with you.

**Reply by devemux86** [reply]
So the summary probably is that regular routing (no beeline):
- Can not enter nogo areas: this [BRouter example](https://brouter.de/brouter-web/#map=12/48.1407/11.5652/standard&lonlats=11.592973,48.086766;11.577211,48.142027&nogos=11.567093,48.141441,5000&profile=car-eco) should not work
- Can exit nogo areas: this [BRouter example](https://brouter.de/brouter-web/#map=12/48.1407/11.5652/standard&lonlats=11.577211,48.142027;11.592973,48.086766&nogos=11.567093,48.141441,5000&profile=car-eco) should work
- What about via / shaping points in nogo areas: [BRouter example](https://brouter.de/brouter-web/#map=12/48.1407/11.5652/standard&lonlats=11.592973,48.086766;11.577211,48.142027;11.504452,48.172466&nogos=11.567093,48.141441,5000&profile=car-eco)
They should be skipped?

---

Better we should have this discussion in the BRouter repository:
- https://github.com/abrensch/brouter/issues/512

**Comment by WolfgangZ42** [comment]
In my opinion all 3 BRouter examples mentioned above should work. If waypoints are set consciously, they should be taken into account - according to the motto: "Trust me - I know what I am doing!"
I don't think we should even need the beeline function for this.
Here again the comparison to "Avoid Highway", where this also works.

**Reply by devemux86** [reply]
Note that in the "Avoid roadblock" function in Cruiser navigation, app sets nogo areas.
If these nogo areas include any via or shaping points, they are automatically skipped.
Otherwise, the selected roadblock avoidance distance could not be maintained.

So the comment by @afischerdev here makes sense?
- https://github.com/abrensch/brouter/issues/512#issuecomment-1474807267

**Comment by WolfgangZ42** [comment]
Okay - learning all the time ... I didn't know that roadblock avoidance also uses nogo areas. **So it makes sense that a nogo area overrules waypoints.** In principle, the discussion about waypoints within the nogo zone was about my motorcycle tours. I am allowed to drive across munich with my motorcycle. 

In the BRouter app, I can obviously suppress the nogo areas:
<img width="200" src="https://user-images.githubusercontent.com/121806238/226163069-479a3838-2b09-4844-adae-3d129a104ac9.png"/>   <img width="200" src="https://user-images.githubusercontent.com/121806238/226163076-ea2f9593-9425-4b98-88a2-5cb29e722212.png"/>   <img width="200" src="https://user-images.githubusercontent.com/121806238/226163086-ade87a01-b741-48c0-b6c7-e46dbcd6f524.png"/>   <img width="200" src="https://user-images.githubusercontent.com/121806238/226163090-da187583-f52d-4d39-9cbb-1025101f24ff.png"/>   

<img width="200" src="https://user-images.githubusercontent.com/121806238/226163094-300f98a9-8707-42ad-b15a-b2a7e342a032.png"/>   <img width="200" src="https://user-images.githubusercontent.com/121806238/226163096-95e6509e-0ea0-465c-97a9-09757f975fd3.png"/>   <img width="200" src="https://user-images.githubusercontent.com/121806238/226163102-6ec053eb-42fd-4a79-a02d-e3b6fc84169a.png"/>   

Maybe in the connection between Cruiser and BRouter (API ?) this can also be done. That would be the perfect solution!

**Reply by devemux86** [reply]
> Maybe in the connection between Cruiser and BRouter (API ?) this can also be done. That would be the perfect solution!

I already have in my list to explore a possible closer connection of Cruiser and BRouter. :slightly_smiling_face:

**Comment by afischerdev** [comment]
@WolfgangZ42 
Thanks for the note.
The screenshots show the procedure to define nogo vetos.
That is: a nogo definition should not be taken for this profile. In your use case, a diesel ban for bicycle does not make much sense.

Only BRouter was incorrect at this point. 
So overall a problem of using way points for nogo area and the game of nogos and vetos.
The new version should fix both.

**Comment by devemux86** [comment]
@WolfgangZ42 

A new version of BRouter (Beta) is available: https://github.com/devemux86/cruiser/discussions/15#discussion-4757829

Can you test the nogo areas? It should work properly now.

**Reply by WolfgangZ42** [reply]
Works like a charm! I tested the radius version - perfect! Although I'll use the polygon area, it's more accurate. 
In case of defining a waypoint in the nogo area now a note pops up and the waypoint will be ignored: 

<img width="200" src="https://user-images.githubusercontent.com/121806238/226743893-61561931-a35a-4e0d-a85a-ce222a31ca20.png"/>   

After reading all arguments in several threads I agree, that this is right way. 
Now I'm looking of a procedure to easily switch the nogo area(s) on and off. 

> That is: a nogo definition should not be taken for this profile.

Would be great, if I can define a nogo veto just for Kurviger-smallroads.brf ...

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the report and the tests!

I plan to offer a user interface to set nogo areas within Cruiser.

**Comment by WolfgangZ42** [comment]
As I said, works very well. Polygon Area defined in > Android > btools.routingapp > brouter > track ...
<img width="200" src="https://user-images.githubusercontent.com/121806238/231090907-1a68ceb7-1bcd-4335-bce2-7de322f32db4.jpg"/>

But it applies to all BRouter profiles (Car and Motorcycle). So when I am driving my motorcycle and I am allowed to drive in the area closed for old diesel vehicles, Cruiser/Brouter refuses it. 

Now I found the solution for this problem (@afischerdev had already hinted that, but I didn't understand) - you can define vetos for certain profiles for individual NoGo Areas:
<img width="200" src="https://user-images.githubusercontent.com/121806238/231091657-8a91d017-9d97-4f85-9c62-2a42138a9324.png"/>    <img width="200" src="https://user-images.githubusercontent.com/121806238/231092241-964be3fd-0b2e-4c74-b6ea-65c2d1793722.png"/>
With this settings I can differentiate between my vehicles just by using separate .pref files, what I do anyway. 
For me the problem is solved - thanks for all tips and hints, even if I have understood them only now by trial and error.

**Reply by devemux86** [reply]
@WolfgangZ42 thanks for the feedback!

I hope we can offer better integration with BRouter in the future and make the process more user friendly.

@afischerdev already knows this and we will see what is possible from each side.

**Reply by WolfgangZ42** [reply]
Before Cruiser 3.0.20, the specifications of the BRouter app were taken over, the custom profiles and also the NoGoAreaVetos. 
Now BRouter is called directly and custom profiles are no longer necessary, because all relevant parameters can be set in Cruiser. The absolutely right strategy!
My BRouter profiles now look like this:
<img width="200" src="https://github.com/devemux86/cruiser/assets/121806238/645175f1-cf35-4ffd-a52e-982c952f3d58"/>

However, I (and probably its only me!) have the previously solved problem with the NoGoArea in Munich again, because no veto is used anymore. Thus, the NoGoArea is always activated, even on foot I could no longer route into the city center. Maybe it is possible that when calling BRouter the veto for individual profiles (for me only motorcar-short is interesting) can be turned off. Would be great!
Currently I solve the problem by renaming the NoGo-gpx-file ... so low priority!

**Reply by devemux86** [reply]
@WolfgangZ42 do you mean that you have put a nogo file in BRouter media folder?

And is this used by all profiles, but you want to only be used by some of the profiles?
Did it work differently before?

**Reply by afischerdev** [reply]
This is a BRouter problem. When using external profiles, the internal profiles and it's parameters are ignored. So the calling app has to transfer the nogo areas. Means no nogo*.gpx in BRouter folder.

**Reply by devemux86** [reply]
@afischerdev thanks for the answer!

Apps have to pass nogo lat,lon,radius via the [service](https://github.com/abrensch/brouter/blob/master/brouter-routing-app/src/main/aidl/btools/routingapp/IBRouterService.aidl), right?

They cannot pass the content of the nogo files?
(as we can pass the content of remote profiles)

**Reply by afischerdev** [reply]
@devemux86 
via service is the way. Please see [aidl file](https://github.com/abrensch/brouter/blob/master/brouter-routing-app/src/main/aidl/btools/routingapp/IBRouterService.aidl).
It has three options: nogos, polylines and polygons.

**Reply by devemux86** [reply]
@afischerdev thanks!

So we will need an interface in Cruiser to define nogo areas (so we can pass them in BRouter).

Info: Cruiser already uses in "Avoid roadblock" navigation action, the nogo functions of BRouter.

**Reply by WolfgangZ42** [reply]
> And is this used by all profiles, but you want to only be used by some of the profiles?
Did it work differently before?

The NoGo file is a polygon in gpx format and resides in android/media/btools.routingapp/brouter/import/tracks
And you can select in the BRouter app to use a veto for it or not per profile - I select no-veto for motorcar-fast 
It worked fine until 3.0.20

> So the calling app has to transfer the nogo areas. Means no nogo*.gpx in BRouter folder.

Could the calling app (Cruiser) put my [nogo_Munich_MR.gpx.txt](https://github.com/devemux86/cruiser/files/11597854/nogo_Munich_MR.gpx.txt) into a specific folder and define "not veto" for motorcar-fast?

**Reply by devemux86** [reply]
> It worked fine until 3.0.20

This is because previous Cruiser versions used the BRouter profiles as they exist in BRouter folder.
This allowed BRouter app to combine its profiles + nogo files.

New Cruiser versions use modified BRouter profiles, so we can offer more BRouter routing settings.
With Cruiser 3.0.21 this only happens with 'car-cruiser' profile, all other profiles are BRouter original.

> Could the calling app (Cruiser) put my [nogo_Munich_MR.gpx.txt](https://github.com/devemux86/cruiser/files/11597854/nogo_Munich_MR.gpx.txt) into a specific folder and define "not veto" for motorcar-fast?

:+1: 
This could be done and I have already put it in my list!

**Comment by devemux86** [comment]
> I can differentiate between my vehicles just by using separate .pref files

Cruiser .pref settings system is extremely powerful to maintain / use different configurations!

In the future I might introduce internal profile settings (like OsmAnd).
Currently different profiles settings exist in:
- Vehicle speeds
- Voice distances

_But this is a discussion for a different topic (if you are interested)._ :slightly_smiling_face:

**Reply by DolomitesTom** [reply]
Sounds interesting, but don't understand what is meant. Don't know Osmand (only Ostmann spices 🙂) and don't have the latest Brouter app since I only use the Google play store for installations.

**Comment by devemux86** [comment]
@WolfgangZ42 

[Cruiser 3.0.22](https://play.google.com/store/apps/details?id=gr.talent.cruiser) reads nogo gpx areas from Cruiser media folder and uses them in BRouter routing.

This is a workaround, we need a better implementation like "Open nogo areas" from any location.

<img src="https://github.com/devemux86/cruiser/assets/3484020/84c7c19a-fa63-4f2e-8fba-b1c6fb58cffb" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/bccacfce-e0d5-4c08-9ad4-b24db3b867cf" width="50%"/>

**Comment by devemux86** [comment]
_"When using external profiles, the internal profiles and it's parameters are ignored. So the calling app has to transfer the nogo areas. **Means no nogo gpx in BRouter folder**."_

_Originally posted by @afischerdev in https://github.com/devemux86/cruiser/discussions/85#discussioncomment-6035477._

---

Please remove all nogo gpx files from BRouter folder: `/Android/media/btools.routingapp/broutes/import/tracks/`
and put them in the Cruiser folder seen above: https://github.com/devemux86/cruiser/discussions/85#discussioncomment-6059725

Until a better solution is implemented that does not require specific folders.
(as we have with the opening of external profiles from anywhere)

**Reply by WolfgangZ42** [reply]
Done:
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/44f8691d-2973-47e2-b489-1fa045ff1593"/> 

And it works:
<img width="500" src="https://github.com/devemux86/cruiser/assets/121806238/876ea717-985f-4216-8d8a-93ed441c6cec"/> 

But: I don't have the ability to use the veto mechanism for my car routing. See: https://github.com/devemux86/cruiser/discussions/85#discussioncomment-6035652
It's really not important, I can control it by renaming the nogo file. Sure it would easier to have a use_veto option in settings.

**Reply by devemux86** [reply]
> Sure it would easier to have a use_veto option in settings.

Something like this is in the works. :slightly_smiling_face: 
I'm currently thinking about the implementation and will have more news soon!

**Comment by devemux86** [comment]
Next Cruiser will be able to open multiple BRouter nogo gpx files from any storage location.

It will be available in Cruiser's advanced settings for BRouter (offline).
You will not have to put anything in BRouter or Cruiser media folders.

Info: there is a special naming convention to specify no-go areas in BRouter:
https://github.com/abrensch/brouter/blob/master/docs/users/android_advanced.md#using-nogo-areas

<img src="https://github.com/devemux86/cruiser/assets/3484020/a02d6f54-e255-4b71-b255-7098a6d76504" width="50%"/><img src="https://github.com/devemux86/cruiser/assets/3484020/0fdedc5b-4ab8-4e5a-a1a9-5e97647cd2f7" width="50%"/>

**Reply by DolomitesTom** [reply]
How can the Nogo area be defined in cruiser without having to read how to create the external file?

**Reply by devemux86** [reply]
> How can the Nogo area be defined in cruiser without having to read how to create the external file?

This will need a user interface and it is already in my implementation list!

Could work with center + radius and should be visible on map.
For example, see the implementation in the [BRouter website](https://brouter.de/brouter-web/).

**Reply by DolomitesTom** [reply]
Wow cool

**Reply by devemux86** [reply, marked as answer]
> Next Cruiser will be able to open multiple BRouter nogo gpx files from any storage location.

> It will be available in Cruiser's advanced settings for BRouter (offline).
> You will not have to put anything in BRouter or Cruiser media folders.

Available in [Cruiser 3.0.23](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

"Settings | Routes | Options | BRouter (offline) | Advanced | BRouter no-go areas".
(you can open multiple gpx files)

**Comment by WolfgangZ42** [comment]
long time ago ... 
My polygon no-go area for Munich (see above) still works perfectly, as do a number of no-go circles for different profiles.

Now I want to define another polygon no-go area for the 95db-ban in Tirol, but I can't seem to do it. Unfortunately, I no longer have any documentation from 2023 and can't remember how I implemented “nogo_Munich_MR.gpx” at the time. 
I have now compiled the new “nogo_Tirol95db.gpx” in the same way as the existing one. I can import and display it, but I can't convert it into a no-go area. 

<img width="200" src="https://github.com/user-attachments/assets/629609d6-a8ed-41a2-bed2-30f5d9f5b2b1"/>...<img width="200" src="https://github.com/user-attachments/assets/1552fa3c-8ae5-4baa-a883-3aee9ca3f1be"/>...<img width="200" src="https://github.com/user-attachments/assets/43c961ae-777b-41a5-948c-0ee496595e3a"/>

Can you please help me?

Here are both files and the folder from my collection: 
[nogo_Munich_MR.gpx.txt](https://github.com/user-attachments/files/23092673/nogo_Munich_MR.gpx.txt)
[nogo_Tirol95db.gpx.txt](https://github.com/user-attachments/files/23092676/nogo_Tirol95db.gpx.txt)
[nogo_areas.zip](https://github.com/user-attachments/files/23092678/nogo_areas.zip)

Btw: I had success by defining several no-go-circles at the end of the "forbidden" roads:

<img width="600" src="https://github.com/user-attachments/assets/68fe8d05-8a04-4adf-97f6-7f31a753c418"/>

 but would prefer a "real no-go-area". 🥴

**Reply by devemux86** [reply]
@WolfgangZ42 No-go areas can be polygons in Tracks (or points with radius in Favorites):

- As a polygon in Tracks, its first and last points must have the same coordinates
(it will have the area icon in the list)

- Set the polygon as "No-go area" using its menu
(the no-go icon is activated) 

- Also it must be set as visible to participate in routing

**Reply by devemux86** [reply]
@WolfgangZ42 The Munich track is closed (first point = last point), so it is considered as a polygon.

The Tirol track is not closed (first point is not the same as last point), so it is considered as a line.

**Reply by WolfgangZ42** [reply]
Sometimes it's quite easy ... 🥴
Thanks a lot Emux!
![2025-10-23 11 39 01](https://github.com/user-attachments/assets/0e52e2cc-7345-4af1-a3c0-b55e9e5744c2)
